[Select] Improve response, react to mouse down

SarthakC opened the pull request:
resolves #8999

eps1lon commented on packages/material-ui/src/Select/SelectInput.js:
```diff
@@ -285,7 +285,7 @@ const SelectInput = React.forwardRef(function SelectInput(props, ref) {
         aria-owns={open ? `menu-${name || ''}` : undefined}
         onKeyDown={handleKeyDown}
-        onClick={disabled || readOnly ? null : handleClick}
+        onMouseDown={disabled || readOnly ? null : handleClick}
```
We want to make sure that any forwarding logic (`onClick` in `handleclick`) is still forwarded properly. So any open logic should be moved into an `handleMouseDown` while the old `handleClick` no longer opens/closes the options.

When we do this we need need to make sure Space and Enter still open the options in their handleClick

SarthakC replied:
Understood. Thanks for taking the time to review this.

eps1lon commented:
I added some tests (https://github.com/SarthakC/material-ui/pull/1) how I think this new behavior should work.

SarthakC commented:
> I added some tests ([SarthakC#1](https://github.com/SarthakC/material-ui/pull/1)) how I think this new behavior should work.

Thank you. I will work on this and revert.

SarthakC commented:
> I added some tests ([SarthakC#1](https://github.com/SarthakC/material-ui/pull/1)) how I think this new behavior should work.

It worked. I added a few more tests based on yours.

eps1lon commented on packages/material-ui/src/Select/Select.test.js:
```diff
@@ -197,7 +197,7 @@ describe('<Select />', () => {
     const { getByRole } = render(<Select value="" autoFocus />);
 
     act(() => {
```
We can remove this now. `fireEvent` uses `act` internally. There are more instances of this in this PR. I think I forgot to remove them. It doesn't have any negative impact but the less code we have the better.

eps1lon commented on packages/material-ui/src/Select/SelectInput.js:
```diff
@@ -321,7 +321,7 @@ const SelectInput = React.forwardRef(function SelectInput(props, ref) {
         aria-haspopup="listbox"
         onKeyDown={handleKeyDown}
-        onClick={disabled || readOnly ? null : handleClick}
+        onMouseDown={disabled || readOnly ? null : handleClick}
```
Could you rename this to `handleMouseDown` and make sure click isn't called from that one?

The tests wouldn't catch this double invokation since they don't simulate all the events that are dispatch when the user clicks with a mouse.

oliviertassinari commented:
@SarthakC Hi. Do you think that you will be able to handle the regressions?

SarthakC commented:
> @SarthakC Hi. Do you think that you will be able to handle the regressions?

@oliviertassinari Hello! Sorry for AWOL. I have been down with the flu. I shall be able to get back to this as soon as I get better.

oliviertassinari commented:
@SarthakC No problem, I hope you will do better :)

oliviertassinari commented:
@SarthakC I have pushed a two-line fixe and rebased with an update of the test. I hope it will do it and that you are doing better.

eps1lon commented on packages/material-ui/src/Select/Select.test.js:
```diff
@@ -93,27 +93,14 @@ describe('<Select />', () => {
     // this means the trigger is blurred immediately
     const handleBlur = spy();
     const { getByRole, getAllByRole, queryByRole } = render(
-      <Select
```
Can we leave this? While it's not necessarily required for the test it simulates how an actual user-click would look like.

oliviertassinari replied:
The select focus on mouse down internally. Do you still want to keep it?

eps1lon replied:
Yes because as I said: This is closer to actual user behavior.

oliviertassinari replied:
Ok, I will revert it :)

oliviertassinari replied:
It sounds like we are on a limit of the test infrastructure. It would be interesting if this could be supported directly. I'm wondering if it's something testing-library has an opportunity with.

eps1lon replied:
It's a bit tricky since this behavior is platform dependent. I just want to avoid unnecessary test changes to be as sure as possible that we understand the new behavior.

SarthakC commented:
> @SarthakC I have pushed a two-line fixe and rebased with an update of the test. I hope it will do it and that you are doing better.

Thank you @oliviertassinari   
I am recovering now. My period of convalescence was longer than I expected. Need to change my city 😅

oliviertassinari commented:
@SarthakC Thank you

herrethan commented:
Wanted to mention we had similar breakage happen due to the Select onClick change from this PR, inside the `renderValue` component:

```jsx
<Select
  renderValue={() => (
    <TextField
      // this no longer fires
      onClick={e => doSomething(e)}
    />
  )}
/>
```
the solution is to now use `onMouseDown` instead of `onClick` here, but would this be recommended for any element you use as the renderValue?
seems like something worth making a little more noise about.

oliviertassinari commented:
@herrethan What's your use case for rendering a TextField here, with on click? 

@eps1lon What do you think of reverting these changes? We have seen a couple of negative effects from this change:
- people using onClick in their unit tests
- people rendering custom elements inside the select

Does it worth it? And if it does, released under a patch, is this right? 🤔

herrethan commented:
@oliviertassinari thanks!

We have a unique input field that is a combo between text input and select dropdown, where the select options are predefined values meant to be convenience selections that populate the text input.

<img src="https://user-images.githubusercontent.com/5460067/70644314-c2b64100-1c10-11ea-8ef6-4610995b1090.png" width="200"/>

On the TextField onClick we `e.stopPropagation()` to allow the user to enter text in the text input. There are surely other ways of achieving this but I was wondering more generally about anyone trying to capture onClick from inside the renderValue.

oliviertassinari commented:
Right, the counter-arguments to revert are that 1. the patch was meant to improve user experience, that the test regression impact unit tests (not e2e), tests that could be less dependent on the event but only consider the user actions. 2. People rendering custom interactive elements in the select, likely want to use https://material-ui.com/components/autocomplete/ instead.

eps1lon commented:
> people using onClick in their unit tests

This would be bad precedent. It was a bug. People relying on this bug (especially in a test) should not block the fix.

> people rendering custom elements inside the select

We have to stop at some point. Customization doesn't mean you should do everything that's possible. 

Especially the example from @herrethan is a classic combobox. It was never intended to use the Select in that way. 

This is why it is important in open source to give back. Have a use case that isn't explicitly documented? Contribute a test case or documentation or open an issue if the use case is not trivial and let maintainers handle the integration. Not sure if it's trivial or not? Open anyway. 

I also want to emphasize that we have to be more confident in our judgement. We can't consider a revert because of a single issue report. Instead figure out how to best move forward.

oliviertassinari commented:
I would say that we had a single report 🙃: #18658, #18655.

eps1lon commented:
> [...] #18655.

I explicitly addressed this one. It was relying on a bug. This should never block a fix. 

The special use case should be considered though.

oliviertassinari commented:
OK thanks. I think that we should leave it more time, to see how things settle. My main concern now (releasing it under a patch, that's too later) is with how people write unit tests, we are making it a bit harder for them but that's DX based and not UX based.

eps1lon commented:
>  we are making it a bit harder for them

No we are fixing a bug. You are now making an argument to never release fixes because people might rely on them in their software.

koustubhmoharir commented:
I use the Select component inside a DataGrid cell. This change broke the logic to enter editing mode in the cell. Attempting to fix it breaks other parts of the DataGrid logic. I will have to spend a few hours getting everything working again. Given that the difference is perceived performance (between click and mouse down) is not even noticeable (at least to me), I wonder if this change was worth it.
